packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,250 @@
+  public readonly name: string;
+
+  constructor(private readonly pattern: any, options: LiteralMatchOptions = {}) {
+    super();
+    this.partialObjects = options.partialObjects ?? false;
+    this.name = 'exact';
+
+    if (IMatcher.isMatcher(this.pattern)) {
+      throw new Error('LiteralMatch cannot directly contain another matcher. ' +
+        'Remove the top-level matcher or nest it more deeply.');
+    }
+  }
+
+  public test(actual: any): MatchResult {
+    if (Array.isArray(this.pattern)) {
+      return new ArrayMatch(this.pattern, { subsequence: false }).test(actual);
+    }
+
+    if (typeof this.pattern === 'object') {
+      return new ObjectMatch(this.pattern, { partial: this.partialObjects }).test(actual);
+    }
+
+    const result = new MatchResult();
+    if (typeof this.pattern !== typeof actual) {
+      result.push(this, [], `Expected type ${typeof this.pattern} but received ${getType(actual)}`);
+      return result;
+    }
+
+    if (this.pattern === ABSENT) {
+      throw new Error('absentProperty() can only be used in an object matcher');
+    }
+
+    if (actual !== this.pattern) {
+      result.push(this, [], `Expected ${this.pattern} but received ${actual}`);
+    }
+
+    return result;
+  }
+}
+
+/**
+ * Options when initializing the `ArrayMatch` class.
+ */
+interface ArrayMatchOptions {
+  /**
+   * Whether the pattern is a subsequence of the target.
+   * A subsequence is a sequence that can be derived from another sequence by deleting
+   * some or no elements without changing the order of the remaining elements.
+   * @default true
+   */
+  readonly subsequence?: boolean;
+}
+
+/**
+ * Match class that matches arrays.
+ */
+class ArrayMatch extends IMatcher {
+  private readonly partial: boolean;
+  public readonly name: string;
+
+  constructor(private readonly pattern: any[], options: ArrayMatchOptions = {}) {
+    super();
+    this.partial = options.subsequence ?? true;
+    if (this.partial) {
+      this.name = 'arrayWith';
+    } else {
+      this.name = 'arrayEquals';
+    }
+  }
+
+  public test(actual: any): MatchResult {
+    if (!Array.isArray(actual)) {
+      return new MatchResult().push(this, [], `Expected type array but received ${getType(actual)}`);
+    }
+    if (!this.partial && this.pattern.length !== actual.length) {
+      return new MatchResult().push(this, [], `Expected array of length ${this.pattern.length} but received ${actual.length}`);
+    }
+
+    let patternIdx = 0;
+    let actualIdx = 0;
+
+    const result = new MatchResult();
+    while (patternIdx < this.pattern.length && actualIdx < actual.length) {
+      const patternElement = this.pattern[patternIdx];
+      const matcher = IMatcher.isMatcher(patternElement) ? patternElement : new LiteralMatch(patternElement);
+      const innerResult = matcher.test(actual[actualIdx]);
+
+      if (!this.partial || !innerResult.hasFailed()) {
+        result.compose(`[${actualIdx}]`, innerResult);
```

Comment:
this is clever but I think it would be more legibile if it's split into two conditionals

Reply:
Also it's wrong. Shouldn't it have been: `(!this.partion || innerResult.hasFailed())` ?

Reply:
It is correct.

Reply:
Ah yeah. It's weird then :D